File: src/mgen/shared/schemas.py
"""Output DataFrame schemas for Data.xlsx sheets.

These define the exact column names and dtypes that the downstream R scripts
expect. Any change here is a breaking change for Phase 2.
"""

from __future__ import annotations

_DATETIME = "datetime64[ns]"

# Macro1 sheet: full-precision metrics, one row per site x date
MACRO1_COLUMNS = ["Site", "Date", "Period", "EPTrich", "EPTabun", "QMCI", "Season"]
MACRO1_DTYPES = {
    "Site": "object",
    "Date": _DATETIME,
    "Period": "object",
    "EPTrich": "float64",
    "EPTabun": "float64",
    "QMCI": "float64",
    "Season": "object",
}

# Macro sheet: aggregated means for replicated sites
MACRO_COLUMNS = MACRO1_COLUMNS
MACRO_DTYPES = MACRO1_DTYPES

# MacroSpecies sheet: long-format taxa tally
MACRO_SPECIES_COLUMNS = [
    "Phase",
    "Date",
    "Site",
    "Taxa",
    "Species",
    "Tally",
    "is_additional",
]
MACRO_SPECIES_DTYPES = {
    "Phase": "object",
    "Date": _DATETIME,
    "Site": "object",
    "Taxa": "object",
    "Species": "object",
    "Tally": "int64",
    "is_additional": "bool",
}

# Sediment sheet: SAM scores
SEDIMENT_COLUMNS = ["Site", "Date", "Period", "SAM1", "SAM3", "Season"]
SEDIMENT_DTYPES = {
    "Site": "object",
    "Date": _DATETIME,
    "Period": "object",
    "SAM1": "float64",
    "SAM3": "float64",
    "Season": "object",
}

# Clarity sheet: water clarity monitoring data.
CLARITY_COLUMNS = [
    "Site",
    "Date",
    "NTU-Fieldmeter",
    "NTU-Continuous Sensor",
    "NTU-Lab",
    "pH-Fieldmeter",
    "pH-Lab",
    "TSS-Lab",
    "Clarity (mm)",
    "Comments",
]
CLARITY_DTYPES = {
    "Site": "object",
    "Date": _DATETIME,
    "NTU-Fieldmeter": "float64",
    "NTU-Continuous Sensor": "float64",
    "NTU-Lab": "float64",
    "pH-Fieldmeter": "float64",
    "pH-Lab": "float64",
    "TSS-Lab": "float64",
    "Clarity (mm)": "float64",
    "Comments": "object",
}

# SedimentSize sheet: grain-size distribution (Wentworth scale).
# The 10 size-class columns pass through with their original source names.
# A future phase may aggregate bins (e.g. split Clay/silt, sum Fines);
# that will require ecologist sign-off and a schema migration.
SEDIMENT_SIZE_COLUMNS = [
    "Site",
    "Date",
    "Period",
    "Season",
    "Clay/silt (<0.06 mm)",
    "Sand (>0.06-2 mm)",
    "Small gravel (>2-8 mm)",
    "Small-med gravel (>8-16 mm)",
    "Med-large gravel (>16-32 mm)",
    "Large gravel (>32-64 mm)",
    "Small cobble (>64-128 mm)",
    "Large cobble (>128-256 mm)",
    "Boulders (>256 mm)",
    "Bedrock",
]
SEDIMENT_SIZE_DTYPES = {
    "Site": "object",
    "Date": _DATETIME,
    "Period": "object",
    "Season": "object",
    **dict.fromkeys(SEDIMENT_SIZE_COLUMNS[4:], "float64"),
}

# RPD sheet: residual pool depth summary per site/date
RPD_COLUMNS = ["Site", "Date", "Count", "Mean", "StdDev", "CI_Lower", "CI_Upper"]
RPD_DTYPES = {
    "Site": "object",
    "Date": _DATETIME,
    "Count": "int64",
    "Mean": "float64",
    "StdDev": "float64",
    "CI_Lower": "float64",
    "CI_Upper": "float64",
}

# LDV sheet: low-flow depth variability (CV%) per site/date
LDV_COLUMNS = ["Site", "Date", "Season", "CV_pct"]
LDV_DTYPES = {
    "Site": "object",
    "Date": _DATETIME,
    "Season": "object",
    "CV_pct": "float64",
}

File: src/r/data/schemas.R
# schemas.R — output column contracts for Data.xlsx sheets.
# Verbatim from src/mgen/shared/schemas.py. Any change here is a breaking
# change for the R figure pipeline.

MACRO1_COLUMNS <- c("Site", "Date", "Period", "EPTrich", "EPTabun", "QMCI", "Season")
MACRO_COLUMNS <- MACRO1_COLUMNS

MACRO_SPECIES_COLUMNS <- c(
  "Phase", "Date", "Site", "Taxa", "Species", "Tally", "is_additional"
)

SEDIMENT_COLUMNS <- c("Site", "Date", "Period", "SAM1", "SAM3", "Season")

CLARITY_COLUMNS <- c(
  "Site", "Date", "NTU-Fieldmeter", "NTU-Continuous Sensor", "NTU-Lab",
  "pH-Fieldmeter", "pH-Lab", "TSS-Lab", "Clarity (mm)", "Comments"
)

SEDIMENT_SIZE_COLUMNS <- c(
  "Site", "Date", "Period", "Season",
  "Clay/silt (<0.06 mm)", "Sand (>0.06-2 mm)", "Small gravel (>2-8 mm)",
  "Small-med gravel (>8-16 mm)", "Med-large gravel (>16-32 mm)",
  "Large gravel (>32-64 mm)", "Small cobble (>64-128 mm)",
  "Large cobble (>128-256 mm)", "Boulders (>256 mm)", "Bedrock"
)

RPD_COLUMNS <- c("Site", "Date", "Count", "Mean", "StdDev", "CI_Lower", "CI_Upper")

LDV_COLUMNS <- c("Site", "Date", "Season", "CV_pct")

# Fish: R-only sheet carrying the raw trapping rows the figure stage needs.
# No Python schemas.py counterpart (the Python pipeline reads fish live).
FISH_COLUMNS <- c(
  "Site", "Catchment", "Date", "Species category (for abundance)", "Number"
)

# MacroSummary: R-only per-site report table (Table 5.6). No Python counterpart.
# One row per Site x Date. Point estimates for single-sample (soft-bottom/D-net)
# sites; mean + 95% CI across replicates for multi-sample (hard-bottom/Surber)
# sites, so the "_CI" half-width columns are NA for single-sample sites.
# Catchment/Substrate/Method are joined from SITE_LEGEND; MCI/QMCI use the
# soft-bottom tolerance variant for soft-bottom sites (see SITES_WITHOUT_REPLICATES).
MACRO_SUMMARY_COLUMNS <- c(
  "Catchment", "Site", "Substrate", "Method", "Date", "Season",
  "NumIndividuals", "NumIndividuals_CI",
  "NumTaxa", "NumTaxa_CI",
  "MCI", "MCI_CI", "MCI_Class",
  "QMCI", "QMCI_CI", "QMCI_Class",
  "PctEPTRichness", "PctEPTRichness_CI",
  "PctEPTAbundance", "PctEPTAbundance_CI",
  "DominantTaxa"
)

# FishSummary: R-only per-site report table (Table 5.8). No Python counterpart.
# One row per Site x Season x Year. Species columns are catch counts (0 when
# absent); Shrimp is a qualitative abundance (Uncommon/Common/Abundant) since
# shrimp are not counted. TotalFish excludes koura and shrimp but includes
# unidentified fish/eels. TaxaRichness counts distinct identified taxa groups
# present (koura counts, shrimp excluded, unidentified not counted separately).
# CPUE is catch (excl. koura & shrimp) per net/trap, separately for mini-fyke
# nets (fykes) and Gee's minnow traps (GMTs). Catchment comes from the source.
FISH_SUMMARY_COLUMNS <- c(
  "Catchment", "Site", "Season", "Year",
  "LongfinEel", "ShortfinEel", "CommonBully", "RedfinBully",
  "BandedKokopu", "GiantKokopu", "Inanga",
  "UnidBully", "UnidKokopu", "UnidEel", "Koura",
  "Shrimp", "TotalFish", "TaxaRichness", "CPUE_fykes", "CPUE_GMTs"
)

SHEET_ORDER <- c(
  "Macro", "Macro1", "MacroSpecies", "Sediment",
  "SedimentSize", "Clarity", "RPD", "LDV", "Fish",
  "MacroSummary", "FishSummary"
)

SCHEMA_MAP <- list(
  Macro = MACRO_COLUMNS,
  Macro1 = MACRO1_COLUMNS,
  MacroSpecies = MACRO_SPECIES_COLUMNS,
  Sediment = SEDIMENT_COLUMNS,
  SedimentSize = SEDIMENT_SIZE_COLUMNS,
  Clarity = CLARITY_COLUMNS,
  RPD = RPD_COLUMNS,
  LDV = LDV_COLUMNS,
  Fish = FISH_COLUMNS,
  MacroSummary = MACRO_SUMMARY_COLUMNS,
  FishSummary = FISH_SUMMARY_COLUMNS
)
